pkg/querier/timeseries_series_set.go

```diff
@@ -76,59 +78,147 @@ func (t *timeseries) Labels() labels.Labels {
+	})
+	t.iH = sort.Search(len(t.ts.series.Histograms), func(i int) bool {
+		return t.ts.series.Histograms[i].Timestamp >= s
+	})
 
-	if t.i < len(t.ts.series.Samples) {
+	if t.iF >= len(t.ts.series.Samples) && t.iH >= len(t.ts.series.Histograms) {
+		return chunkenc.ValNone
+	}
+	if t.iF >= len(t.ts.series.Samples) {
+		t.atH = true
+		return t.atTypeHisto()
+	}
+	if t.iH >= len(t.ts.series.Histograms) {
+		t.atH = false
+		return chunkenc.ValFloat
+	}
+	if t.ts.series.Samples[t.iF].TimestampMs < t.ts.series.Histograms[t.iH].Timestamp {
+		t.iH--
+		t.atH = false
 		return chunkenc.ValFloat
 	}
-	return chunkenc.ValNone
+	t.iF--
+	t.atH = true
+	return t.atTypeHisto()
 }
 
 // At implements the implements chunkenc.Iterator.
 func (t *timeSeriesSeriesIterator) At() (int64, float64) {
-	if t.i < 0 || t.i >= len(t.ts.series.Samples) {
+	if t.atH {
+		panic(errors.New("timeSeriesSeriesIterator: Calling At() when cursor is at histogram"))
+	}
+	if t.iF < 0 || t.iF >= len(t.ts.series.Samples) {
 		return 0, 0
 	}
-	return t.ts.series.Samples[t.i].TimestampMs, t.ts.series.Samples[t.i].Value
+	return t.ts.series.Samples[t.iF].TimestampMs, t.ts.series.Samples[t.iF].Value
 }
 
 // AtHistogram implements chunkenc.Iterator.
 func (t *timeSeriesSeriesIterator) AtHistogram() (int64, *histogram.Histogram) {
-	panic(errors.New("timeSeriesSeriesIterator: AtFloatHistogram is not implemented"))
+	if !t.atH {
+		panic(errors.New("timeSeriesSeriesIterator: Calling AtHistogram() when cursor is not at histogram"))
+	}
+	if t.iH < 0 || t.iH >= len(t.ts.series.Histograms) {
+		return 0, nil
+	}
+	h := t.ts.series.Histograms[t.iH]
+	return h.Timestamp, mimirpb.FromHistogramProtoToHistogram(&h)
 }
 
 // AtFloatHistogram implements chunkenc.Iterator.
 func (t *timeSeriesSeriesIterator) AtFloatHistogram() (int64, *histogram.FloatHistogram) {
-	panic(errors.New("timeSeriesSeriesIterator: AtFloatHistogram is not implemented"))
+	if !t.atH {
+		panic(errors.New("timeSeriesSeriesIterator: Calling AtFloatHistogram() when cursor is not at histogram"))
+	}
+	if t.iH < 0 || t.iH >= len(t.ts.series.Histograms) {
+		return 0, nil
+	}
+	h := t.ts.series.Histograms[t.iH]
+	// Should we automatially convert to float if the hisotgram happens to be integer histogram?
+	return h.Timestamp, mimirpb.FromHistogramProtoToFloatHistogram(&h)
 }
 
 // AtT implements implements chunkenc.Iterator.
 func (t *timeSeriesSeriesIterator) AtT() int64 {
-	ts, _ := t.At()
-	return ts
+	if t.atH {
+		if t.iH < 0 || t.iH >= len(t.ts.series.Histograms) {
+			return 0
+		}
+		return t.ts.series.Histograms[t.iH].Timestamp
+	}
+	if t.iF < 0 || t.iF >= len(t.ts.series.Samples) {
+		return 0
+	}
+	return t.ts.series.Samples[t.iF].TimestampMs
 }
 
 // Next implements implements chunkenc.Iterator.
 func (t *timeSeriesSeriesIterator) Next() chunkenc.ValueType {
-	t.i++
-	if t.i < len(t.ts.series.Samples) {
+	if t.iF+1 >= len(t.ts.series.Samples) && t.iH+1 >= len(t.ts.series.Histograms) {
+		t.iF++
+		t.iH++
```

Comment:
Should be set to length of those slices instead of increment to avoid overflows

Reply:
fixed